### PR TITLE
Support for GatherND layer

### PR DESCRIPTION
**Pull request by Abdurrahheem**
This PR adds support for GatherND layer. The layer was in comformance deny list initially. 

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

File: modules/dnn/src/layers/gatherND.cpp [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+        indices_strides[q - 1] = 1;
+        for (int i = q - 2; i >= 0; --i)
+            indices_strides[i] = indices_strides[i + 1] * indices.size[i + 1];
+
+        const int outer_size = indices.total() / last_indices_dim;
+        const int inner_size = out.total() / outer_size;
+
+        for (size_t i = 0; i < outer_size; ++i)
+        {
+            std::vector<int> sliced_indices(indices_ptr + i * last_indices_dim, indices_ptr + (i + 1) * last_indices_dim);
```
Comment (asmorkalov) [inline modules/dnn/src/layers/gatherND.cpp:148]: you do not need vector here. just:
```
int* sliced_indices = indices_ptr + i * last_indices_dim;
```

File: modules/dnn/src/layers/gatherND.cpp [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+                switch (dtype) {
+                    case CV_8U: forward_impl<int32_t, uchar>(data, indices, out); break;
+                    case CV_8S: forward_impl<int32_t, schar>(data, indices, out); break;
+                    case CV_32S: forward_impl<int32_t, int32_t>(data, indices, out); break;
+                    case CV_16F: forward_impl<int32_t, float16_t>(data, indices, out); break;
+                    case CV_32F: forward_impl<int32_t, float>(data, indices, out); break;
+                    case CV_64F: forward_impl<int32_t, double>(data, indices, out); break;
+                    default: CV_Error(Error::StsNotImplemented, "Unsupported data type");
+                }
+            }
```
Comment (asmorkalov) [inline modules/dnn/src/layers/gatherND.cpp:104]: `break` is missing:
```
2024-09-03T16:01:13.0386984Z /home/ci/opencv/modules/dnn/src/layers/gatherND.cpp:95:17: warning: this statement may fall through [-Wimplicit-fallthrough=]
2024-09-03T16:01:13.0388547Z    95 |                 switch (dtype) {
2024-09-03T16:01:13.0389213Z       |                 ^~~~~~
2024-09-03T16:01:13.0390112Z /home/ci/opencv/modules/dnn/src/layers/gatherND.cpp:105:13: note: here
2024-09-03T16:01:13.0391087Z   105 |             case CV_64S:
2024-09-03T16:01:13.0391668Z       |             ^~~~
```

File: modules/dnn/src/layers/gatherND.cpp [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+        int itype = indices.depth();
+
+        switch (itype) {
+            case CV_32S:
+            {
+                switch (dtype) {
+                    case CV_8U: forward_impl<int32_t, uchar>(data, indices, out); break;
+                    case CV_8S: forward_impl<int32_t, schar>(data, indices, out); break;
+                    case CV_32S: forward_impl<int32_t, int32_t>(data, indices, out); break;
+                    case CV_16F: forward_impl<int32_t, float16_t>(data, indices, out); break;
```
Comment (asmorkalov) [inline modules/dnn/src/layers/gatherND.cpp:99]:
```
/home/ci/opencv/modules/dnn/src/layers/gatherND.cpp:99:56: error: 'float16_t' was not declared in this scope; did you mean 'float_t'?
2024-09-03T15:59:18.3161177Z    99 |                     case CV_16F: forward_impl<int32_t, float16_t>(data, indices, out); break;
```

Comment (asmorkalov) [reply modules/dnn/src/layers/gatherND.cpp:99]: I propose to use s16 branch for `CV_16F` case to be hardware independent.

File: modules/dnn/src/layers/gatherND.cpp [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+
+        std::vector<int> indices_strides(q);
+        indices_strides[q - 1] = 1;
+        for (int i = q - 2; i >= 0; --i)
+            indices_strides[i] = indices_strides[i + 1] * indices.size[i + 1];
+
+        const int outer_size = indices.total() / last_indices_dim;
+        const int inner_size = out.total() / outer_size;
+
+        for (size_t i = 0; i < outer_size; ++i)
```
Comment (fengyuentau) [inline modules/dnn/src/layers/gatherND.cpp:152]: Looks like it can be parallelized

File: modules/dnn/src/layers/gatherND.cpp [hunk 5]
```diff
@@ -96,24 +97,25 @@ class GatherNDLayerImpl CV_FINAL : public GatherNDLayer
                     case CV_8U: forward_impl<int32_t, uchar>(data, indices, out); break;
                     case CV_8S: forward_impl<int32_t, schar>(data, indices, out); break;
                     case CV_32S: forward_impl<int32_t, int32_t>(data, indices, out); break;
-                    case CV_16F: forward_impl<int32_t, float16_t>(data, indices, out); break;
+                    case CV_16F: forward_impl<int32_t, short>(data, indices, out); break;
```
Comment (asmorkalov) [inline modules/dnn/src/layers/gatherND.cpp:100]: int16_t is better here to ensure size.

File: modules/dnn/src/layers/gatherND.cpp [hunk 6]
```diff
@@ -96,24 +97,25 @@ class GatherNDLayerImpl CV_FINAL : public GatherNDLayer
...
-            }
+            } break;
             case CV_64S:
             {
                 switch (dtype) {
                     case CV_8U: forward_impl<int64_t, uchar>(data, indices, out); break;
                     case CV_8S: forward_impl<int64_t, schar>(data, indices, out); break;
                     case CV_32S: forward_impl<int64_t, int32_t>(data, indices, out); break;
-                    case CV_16F: forward_impl<int64_t, float16_t>(data, indices, out); break;
+                    case CV_16F: forward_impl<int64_t, short>(data, indices, out); break;
```
Comment (asmorkalov) [inline modules/dnn/src/layers/gatherND.cpp:112]: The same.

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 7]
```diff
@@ -3213,6 +3214,15 @@ void ONNXImporter::parseHardmax(LayerParams& layerParams, const opencv_onnx::Nod
     addLayer(layerParams, node_proto);
 }
 
+void ONNXImporter::parseGatherND(LayerParams& layerParams, const opencv_onnx::NodeProto& node_proto)
```
[the remark this answers is not in the document]
Comment (Abdurrahheem) [reply modules/dnn/src/onnx/onnx_importer.cpp:3217]: I propose to leave this as it is, for new parser it makes sence

File: modules/dnn/src/layers/gatherND.cpp (hunk 1 above)
Comment (Abdurrahheem) [reply modules/dnn/src/layers/gatherND.cpp]: This leads to crush during build time

Comment (asmorkalov) [reply modules/dnn/src/layers/gatherND.cpp]: I just pushed fix. `indices_ptr` has template type `iT*`  and `sliced_indices` should. It's not always `int`. The solution with vector just hides the problem.

**Review by asmorkalov — APPROVED** [review]

**Comment by asmorkalov** [comment]
@fengyuentau @dkurt Please take a look. Let's merge, if you do not have objections.

**Review by fengyuentau — APPROVED** [review]

**Pull request merged**